### PR TITLE
Int|null variables must be integer or null

### LINKED ISSUE #34958: [Issue] Int|null variables must be integer or null
This issue is automatically created based on existing pull request: magento/magento2#34140: Int|null variables must be integer or null

---------
null

### PR DESCRIPTION
**Pull request by sheepfy**
null

### Resolved issues:
1. [x] resolves magento/magento2#34958: Int|null variables must be integer or null

### CODE REVIEW

File: app/code/Magento/Catalog/Model/ProductRepository.php [hunk 1]
```diff
@@ -280,6 +280,7 @@ public function __construct(
      */
     public function get($sku, $editMode = false, $storeId = null, $forceReload = false)
     {
+        $storeId = $storeId === null ? null : (int) $storeId;
```
Comment (orlangur) [inline app/code/Magento/Catalog/Model/ProductRepository.php]: This change does not make sense. Numeric strings are also perfectly valid here.

Did you face with any concrete bug?

**Review by orlangur — CHANGES_REQUESTED** [review]

File: app/code/Magento/Catalog/Model/ProductRepository.php (hunk 1 above)
Comment (sheepfy) [reply app/code/Magento/Catalog/Model/ProductRepository.php]: For example:
vendor/magento/module-advanced-checkout/Controller/Cart/UpdateFailedItemOptions.php -> $this->_objectManager->get(\Magento\Store\Model\StoreManager::class)->getStore()->getId() this one returns string


$this->product = $this->productRepository->getById($entityId, true, $this->getStore()->getId()); -> this line here -> StagingLocator.php

etc

Comment (sheepfy) [reply app/code/Magento/Catalog/Model/ProductRepository.php]: Numeric strings are allowed, yes. 
But in the same process, if you call repo->getByIf(.., '1', ..) and repo->getByIf(.., 1, ..) you will load product twice.

Comment (sheepfy) [reply app/code/Magento/Catalog/Model/ProductRepository.php]: because cacheKey will be different

Comment (orlangur) [reply app/code/Magento/Catalog/Model/ProductRepository.php]: Okej, nice catch, does actually make sense with such explanation :) Will review deeper a bit later, need to check this file for other occurrences.

Comment (sheepfy) [reply app/code/Magento/Catalog/Model/ProductRepository.php]: Thank you!

Comment (Den4ik) [reply app/code/Magento/Catalog/Model/ProductRepository.php]: @sheepfy You opinion make sense. But I believe that `$storeId` type casting will be more safe on cache key calculation

```
$cacheKey = $this->getCacheKey([$editMode, $storeId === null ? $storeId : (int)$storeId]);
```

**Review by Den4ik — CHANGES_REQUESTED** [review]

File: app/code/Magento/Catalog/Model/ProductRepository.php (hunk 1 above)
Comment (sheepfy) [reply app/code/Magento/Catalog/Model/ProductRepository.php]: @Den4ik  changes applied

**Comment by Den4ik** [comment]
@magento run all tests

File: app/code/Magento/Catalog/Model/ProductRepository.php (hunk 1 above)
Comment (Den4ik) [reply app/code/Magento/Catalog/Model/ProductRepository.php]: @sheepfy Thanks for update. I'll check it today

**Review by Den4ik — CHANGES_REQUESTED** [review]
@sheepfy Please look at failed static tests

File: app/code/Magento/Catalog/Model/ProductRepository.php (hunk 1 above)
Comment (sheepfy) [reply app/code/Magento/Catalog/Model/ProductRepository.php]: @Den4ik you mean these?

```PHP Code Mess has found error(s):
/var/www/html/app/code/Magento/Catalog/Model/ProductRepository.php:42	The class ProductRepository has an overall complexity of 112 which is very high. The configured complexity threshold is 100.
/var/www/html/app/code/Magento/Catalog/Model/ProductRepository.php:526	The method save() has 115 lines of code. Current threshold is set to 100. Avoid really long methods.```

File: app/code/Magento/Catalog/Model/ProductRepository.php [hunk 2]
```diff
@@ -295,7 +295,7 @@ public function get($sku, $editMode = false, $storeId = null, $forceReload = fal
                 $product->setData('_edit_mode', true);
             }
             if ($storeId !== null) {
-                $product->setData('store_id', $storeId);
+                $product->setData('store_id', (int) $storeId);
```
Comment (Den4ik) [inline app/code/Magento/Catalog/Model/ProductRepository.php:299]: `$storeId` could be `null` and should not be casted here

**Review by Den4ik — CHANGES_REQUESTED** [review]

File: app/code/Magento/Catalog/Model/ProductRepository.php (hunk 2 above)
Comment (sheepfy) [reply app/code/Magento/Catalog/Model/ProductRepository.php:299]: It cannot be null at that line.

Comment (sheepfy) [reply app/code/Magento/Catalog/Model/ProductRepository.php:299]: is just how if statements works

Comment (Den4ik) [reply app/code/Magento/Catalog/Model/ProductRepository.php:299]: Oh, yep. I missed `if check`

File: app/code/Magento/Catalog/Model/ProductRepository.php (hunk 1 above)
Comment (Den4ik) [reply app/code/Magento/Catalog/Model/ProductRepository.php]: @sheepfy You could add suppress warnings for this issue and we were go with this PR

**Review by sidolov — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sidolov, thank you for the review.
[ENGCOM-9330](https://jira.corp.magento.com/browse/ENGCOM-9330) has been created to process this Pull Request
:eight_spoked_asterisk: @sidolov, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Review by Den4ik — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @Den4ik, thank you for the review.
[ENGCOM-9330](https://jira.corp.magento.com/browse/ENGCOM-9330) has been created to process this Pull Request
:eight_spoked_asterisk: @Den4ik, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Alfa** [comment]
Hi @sheepf, @Den4ik and @orlangur Thanks for your collaboration and contribution!

There is no steps to reproduce in description section, still deeply looked into the file changes and review comments because file changes seem to be valid. Below is our observation.

1. As per doc block $storeId should be int or null. Tried reproducing by giving _$storeId_ as string (eg: $storeId = "1") inside the function body i.e. `app/code/Magento/Catalog/Model/ProductRepository.php::public function get($sku, $editMode = false, $storeId = null, $forceReload = false)`. But we are unable to notice the any exceptions/error messages as such when call is made to the same get method through API request. We agree _$cacheKey_ will be different depending on storeId whether is string or int or null, anyhow we get key in return.
    **So, as we are unable to reproduce the issue prior accepting the changes, could you please help us with detailed steps to reproduce in-order to find the exception/ error for the existing code**

2. Note: Proposed changes can also be applied to another function _getById_ available in the same class, can we know reason for applying the changes to only one function.

**Comment by engcom-Alfa** [comment]
@magento Functional Tests EE, Static Tests

**Comment by engcom-Alfa** [comment]
@magento run Functional Tests EE, Static Tests

**Comment by engcom-Alfa** [comment]
May I please get reply to my [above-comment](https://github.com/magento/magento2/pull/34140#issuecomment-1000366635) possibly, so that I can progress the ticket please! @sheepfy
Thanks in advance!

**Comment by sheepfy** [comment]
@engcom-Alfa 
...........
omg. Just make a controller, /whatever/index/index .

public function execute() {
     $sku = 'whatever';
     $this->repo->get($sku, false, 1);
     $this->repo->get($sku, false, '1');
}

Notice that all the mysql query are doubled because of different cache key.


Anyway, you can drop this PR..

**Comment by sheepfy** [comment]
@engcom-Alfa 
also, no one said anything about any exception on API requests.

**Comment by engcom-Alfa** [comment]
@magento run Functional Tests EE, Static Tests

**Comment by engcom-Alfa** [comment]
Thanks for the valuable information @sheepfy It helped me to understand the ticket better and progress

**Comment by engcom-Alfa** [comment]
Thanks for the valuable information @sheepfy It helped me to understand the ticket better and progress

**Comment by engcom-Alfa** [comment]
> @engcom-Alfa also, no one said anything about any exception on API requests.

Thanks for the valuable information @sheepfy It helped me to understand the ticket better and progress

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

1. Have Magento Installed with sample data


**Manual testing scenario:**

1. Goto Magento Catalog module's ProductReporsitory.php class and alter the code as shown in below screenshot
Screenshot 2022-01-06 at 8 43 29 PM

2. Get the admin token from APi request and run the below shown screenshot's API request
image
 
3. Validate the responded cacheKeys

4. Do the PR changes and again hit the same above API and validate the responded keys


**Before:** :heavy_multiplication_x: 2 different cache keys were retrieved for one store id

Screenshot 2022-01-06 at 8 37 14 PM


**After:** :heavy_check_mark: One cache key has been retrieved for one store id

Screenshot 2022-01-06 at 8 45 17 PM


There is no additional regression is required as part of testing since it is relevant to the improvement in the performance.

**Comment by engcom-Alfa** [comment]
@magento run Static Tests

**Comment by engcom-Alfa** [comment]
@magento run Static Tests

**Comment by engcom-Lima** [comment]
Seems like bot moved it to Ready for Testing column automatically, hence moved it back in Extended Testing

**Comment by engcom-Lima** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
@magento run Static Tests

**Comment by engcom-Lima** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
@magento run Functional Tests B2B,Functional Tests CE,Functional Tests EE

**Comment by engcom-Lima** [comment]
@magento run Functional Tests B2B,Functional Tests CE,Functional Tests EE

**Comment by engcom-Lima** [comment]
@magento run Functional Tests CE

**Comment by engcom-Lima** [comment]
@magento run Functional Tests CE

**Comment by engcom-Lima** [comment]
@magento run Functional Tests CE

**Comment by engcom-Lima** [comment]
@magento run Functional Tests CE

**Comment by engcom-Lima** [comment]
@magento run Functional Tests CE

**Comment by engcom-Lima** [comment]
@magento run Functional Tests CE

**Comment by engcom-Lima** [comment]
@magento run Functional Tests CE

**Comment by engcom-Lima** [comment]
@magento run Functional Tests CE

**Comment by engcom-Lima** [comment]
@engcom-Alfa the builds are passed now, moving it to ready for testing

**Comment by engcom-Alfa** [comment]
@magento create issue

**Comment by engcom-Alfa** [comment]
As mentioned in the [above-comment](https://github.com/magento/magento2/pull/34140#issuecomment-1006672844), the PR is Qa passed. Also we have all the [auto-tests passed now](https://github.com/magento/magento2/pull/34140#issuecomment-1010943879), thanks @engcom-Lima for that. Hence moving the PR for _Merge In Progress_.

**Pull request merged**